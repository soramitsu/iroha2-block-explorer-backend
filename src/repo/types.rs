use crate::repo::util::AsText;
use crate::schema;
use crate::schema::TransactionStatus;
use chrono::{DateTime, Utc};
use iroha::data_model::{account, asset, domain, metadata, prelude};
use serde::Deserialize;
use sqlx::types::Json;
use sqlx::{FromRow, Type};
use std::fmt::{Display, Formatter};
use std::num::NonZero;
use std::str::FromStr;

#[derive(Debug, FromRow)]
pub struct Domain {
    pub id: DomainId,
    pub logo: Option<IpfsPath>,
    pub metadata: Metadata,
    pub owned_by: AccountId,
    pub accounts: u32,
    pub assets: u32,
    pub nfts: u32,
}

#[derive(Debug, Type)]
#[sqlx(transparent)]
pub struct DomainId(pub AsText<domain::DomainId>);

#[derive(Debug, Type)]
#[sqlx(transparent)]
pub struct AccountId(pub AsText<account::AccountId>);

#[derive(Debug, Type)]
#[sqlx(transparent)]
pub struct AssetDefinitionId(pub AsText<asset::AssetDefinitionId>);

#[derive(Debug, Type)]
#[sqlx(transparent)]
pub struct AssetId(pub AsText<prelude::AssetId>);

#[derive(Debug, Type)]
#[sqlx(transparent)]
pub struct NftId(pub AsText<prelude::NftId>);

#[derive(Debug, Type)]
#[sqlx(transparent)]
pub struct IpfsPath(pub String);

#[derive(Debug, Type, Deserialize)]
#[sqlx(transparent)]
pub struct Metadata(pub Option<Json<metadata::Metadata>>);

impl From<Metadata> for metadata::Metadata {
    fn from(value: Metadata) -> Self {
        value.0.map(|x| x.0).unwrap_or_default()
    }
}

#[derive(Debug, FromRow)]
pub struct Block {
    pub hash: Hash,
    pub height: NonZero<u64>,
    pub prev_block_hash: Option<Hash>,
    /// None for an empty block
    pub transactions_hash: Option<Hash>,
    pub created_at: DateTime<Utc>,
    pub transactions_total: u32,
    pub transactions_rejected: u32,
}

#[derive(Debug, Type)]
#[sqlx(transparent)]
pub struct Hash(pub AsText<iroha::crypto::Hash>);

#[derive(Debug, Type)]
#[sqlx(transparent)]
pub struct Signature(pub AsText<SignatureDisplay>);

// FIXME: remove when Iroha Signature impls FromStr
#[derive(Debug)]
pub struct SignatureDisplay(pub iroha::crypto::Signature);

impl FromStr for SignatureDisplay {
    type Err = iroha::crypto::error::ParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(Self(iroha::crypto::Signature::from_hex(s)?))
    }
}

impl Display for SignatureDisplay {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        // FIXME: extract hex directly, not via serde_json
        let serde_json::Value::String(value) = serde_json::to_value(&self.0)
            .expect("nothing could fail during signature serialisation to JSON")
        else {
            unreachable!("should always be a string")
        };
        write!(f, "{value}")
    }
}

#[derive(Debug, Type)]
pub enum Executable {
    Instructions,
    #[allow(clippy::upper_case_acronyms)]
    WASM,
}

#[derive(Debug, FromRow)]
pub struct TransactionBase {
    pub hash: Hash,
    pub block: u32,
    pub created_at: DateTime<Utc>,
    pub authority: AccountId,
    pub executable: Executable,
    pub status: TransactionStatus,
}

#[derive(Debug, FromRow)]
pub struct TransactionDetailed {
    #[sqlx(flatten)]
    pub base: TransactionBase,
    pub nonce: Option<NonZero<u32>>,
    pub metadata: Metadata,
    pub time_to_live_ms: u64,
    pub signature: Signature,
    pub rejection_reason: Option<Json<prelude::TransactionRejectionReason>>,
}

#[derive(Debug, FromRow)]
pub struct Account {
    pub id: AccountId,
    pub metadata: Metadata,
    pub owned_domains: u32,
    pub owned_assets: u32,
    pub owned_nfts: u32,
}

#[derive(Debug, FromRow)]
pub struct AssetDefinition {
    pub id: AssetDefinitionId,
    pub owned_by: AccountId,
    pub logo: Option<IpfsPath>,
    pub metadata: Metadata,
    pub mintable: Mintable,
    pub assets: u32,
}

#[derive(Debug, Type)]
pub enum Mintable {
    Infinitely,
    Once,
    Not,
}

#[derive(Debug, FromRow)]
pub struct Asset {
    pub id: AssetId,
    pub value: Json<prelude::Numeric>,
}

#[derive(Debug, FromRow)]
pub struct Nft {
    pub id: NftId,
    pub owned_by: AccountId,
    pub content: Metadata,
}

#[derive(Debug, FromRow)]
pub struct Instruction {
    pub transaction_hash: Hash,
    pub block: u32,
    pub transaction_status: TransactionStatus,
    pub created_at: DateTime<Utc>,
    pub kind: schema::InstructionKind,
    pub r#box: Json<prelude::InstructionBox>,
    pub authority: AccountId,
}
