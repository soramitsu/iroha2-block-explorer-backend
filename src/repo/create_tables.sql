create table if not exists blocks
(
    height            integer primary key not null,
    hash              text                not null,
    prev_block_hash   text,
    transactions_hash text,
    created_at        datetime            not null
);

create table if not exists domains
(
    name     text primary key not null,
    logo     text,
    metadata json
);

create table if not exists accounts
(
    signatory text not null,
    domain    text not null references domains (name),
    metadata  json,
    primary key (signatory, domain)
);

create table if not exists domain_owners
(
    account_signatory text not null,
    account_domain    text not null,
    domain            text not null references domains (name),
    foreign key (account_signatory, account_domain) references accounts (signatory, domain),
    primary key (account_signatory, account_domain, domain)
);

create table if not exists asset_definitions
(
    name               text                                                   not null,
    domain             text                                                   not null references domains (name),
    owned_by_signatory text                                                   not null,
    owned_by_domain    text                                                   not null,
    logo               text,
    metadata           json,
    mintable           text check (mintable in ('Once', 'Not', 'Infinitely')) not null,
    primary key (name, domain),
    foreign key (owned_by_signatory, owned_by_domain) references accounts (signatory, domain)
);


create table if not exists nfts
(
    name               text not null,
    domain             text not null references domains (name),
    owned_by_signatory text not null,
    owned_by_domain    text not null,
    content            json,
    primary key (name, domain),
    foreign key (owned_by_signatory, owned_by_domain) references accounts (signatory, domain)
);

create table if not exists assets
(
    definition_name    text not null,
    definition_domain  text not null,
    owned_by_signatory text not null,
    owned_by_domain    text not null,
    value              json not null,
    foreign key (definition_name, definition_domain) references asset_definitions (name, domain),
    foreign key (owned_by_signatory, owned_by_domain) references accounts (signatory, domain),
    primary key (definition_name, definition_domain, owned_by_signatory, owned_by_domain)
);

create table if not exists transactions
(
    hash                text primary key                                    not null,
    block               integer                                             not null references blocks (height),
    created_at          datetime                                            not null,
    authority_signatory text                                                not null,
    authority_domain    text                                                not null,
    signature           text                                                not null,
    nonce               integer,
    metadata            json,
    time_to_live_ms     integer,
    executable          text check (executable in ('Instructions', 'WASM')) not null,
    error               json,
    foreign key (authority_signatory, authority_domain) references accounts (signatory, domain)
);

create table if not exists instructions
(
    transaction_hash text not null references transactions (hash),
    value            json not null
);

/* VIEWS */

create view if not exists v_transactions
as
select *,
       format('%s@%s', authority_signatory, authority_domain)       as authority,
       case when error is null then 'committed' else 'rejected' end as status
from transactions;

create view if not exists v_instructions
as
select json_each.key                                                as kind,
       instructions.value                                           as box,
--        case
--            /* TODO: truncate payload for `Upgrade` instruction kind? */
--            when json_each.type in ('null', 'text', 'integer', 'real') then json_quote(json_each.value)
--            when json_each.type in ('false', 'true') then json_each.type
--            else json_each.value
--            end                                                      as payload,
       created_at,
       transaction_hash,
       case when error is null then 'committed' else 'rejected' end as transaction_status,
       format('%s@%s', authority_signatory, authority_domain)       as authority,
       block
from instructions,
     json_each(instructions.value)
         join v_transactions as txs on txs.hash = instructions.transaction_hash;

create view if not exists v_assets as
select *,
       case assets.definition_domain = assets.owned_by_domain
           when true then format('%s##%s@%s', assets.definition_name, assets.owned_by_signatory, assets.owned_by_domain)
           else format('%s#%s#%s@%s', assets.definition_name, assets.definition_domain, assets.owned_by_signatory,
                       assets.owned_by_domain) end as id,
       value
from assets;

create view if not exists v_nfts as
select *,
       format('%s$%s', nfts.name, nfts.domain)                        as id,
       format('%s@%s', nfts.owned_by_signatory, nfts.owned_by_domain) as owned_by
from nfts;
