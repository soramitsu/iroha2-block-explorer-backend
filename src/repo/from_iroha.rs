use crate::repo::{AsText, SignatureDisplay};
use chrono::DateTime;
use eyre::Result;
use iroha::client::Client;
use iroha::data_model::prelude::*;
use serde_json::json;
use sqlx::types::Json;
use sqlx::{query, QueryBuilder, SqliteConnection};
use tokio::task::spawn_blocking;
use tracing::debug;

/// Scan Iroha into an `SQLite` database.
#[allow(clippy::too_many_lines)]
pub async fn scan_into(client: &Client, conn: &mut SqliteConnection) -> Result<()> {
    debug!("Scanning Iroha into an in-memory SQLite database");

    debug!("Fetching data from Iroha...");
    let client = client.clone();
    let (domains, accounts, blocks, assets_definitions, assets) = spawn_blocking(move || {
        let domains = client.query(FindDomains).execute_all()?;
        let accounts = client.query(FindAccounts).execute_all()?;
        let blocks = client.query(FindBlocks).execute_all()?;
        let assets_definitions = client.query(FindAssetsDefinitions).execute_all()?;
        let assets = client.query(FindAssets).execute_all()?;
        Ok::<_, eyre::Report>((domains, accounts, blocks, assets_definitions, assets))
    })
    .await??;
    debug!("Done fetching");

    query(include_str!("./create_tables.sql"))
        .execute(&mut *conn)
        .await?;
    query("PRAGMA foreign_keys=OFF").execute(&mut *conn).await?;
    query("BEGIN TRANSACTION").execute(&mut *conn).await?;

    // todo: handle empty data
    debug!("Inserting domains & accounts...");
    QueryBuilder::new("insert into domains(name, logo, metadata) ")
        .push_values(&domains, |mut b, value| {
            b.push_bind(value.id().name().as_ref())
                .push_bind(value.logo().as_ref().map(AsText))
                .push_bind(Json(value.metadata()));
        })
        .build()
        .execute(&mut *conn)
        .await?;
    QueryBuilder::new("insert into accounts(signatory, domain, metadata) ")
        .push_values(&accounts, |mut b, value| {
            b.push_bind(AsText(value.signatory()))
                .push_bind(AsText(value.id().domain()))
                .push_bind(Json(value.metadata()));
        })
        .build()
        .execute(&mut *conn)
        .await?;
    QueryBuilder::new("insert into domain_owners(account_signatory, account_domain, domain) ")
        .push_values(&domains, |mut b, value| {
            b.push_bind(AsText(value.owned_by().signatory()))
                .push_bind(AsText(value.owned_by().domain()))
                .push_bind(AsText(value.id()));
        })
        .build()
        .execute(&mut *conn)
        .await?;

    // TODO: handle empty blocks, txs, isis
    debug!("Inserting blocks, transactions, instructions...");
    let mut b = QueryBuilder::new("insert into blocks(");
    let mut sep = b.separated(", ");
    for i in [
        "hash",
        "height",
        "created_at",
        "prev_block_hash",
        "transactions_hash",
    ] {
        sep.push(i);
    }
    b.push(") ")
        .push_values(&blocks, |mut b, value| {
            b.push_bind(AsText(value.hash()))
                .push_bind(value.header().height().get() as i32)
                .push_bind(DateTime::from_timestamp_millis(
                    value.header().creation_time().as_millis() as i64,
                ))
                .push_bind(value.header().prev_block_hash().map(AsText))
                .push_bind(AsText(value.header().transactions_hash()));
        })
        .build()
        .execute(&mut *conn)
        .await?;

    let mut b = QueryBuilder::new("insert into transactions(");
    let mut sep = b.separated(", ");
    for i in &[
        "hash",
        "block",
        "created_at",
        "time_to_live_ms",
        "authority_signatory",
        "authority_domain",
        "signature",
        "nonce",
        "metadata",
        "error",
        "executable",
    ] {
        sep.push(i);
    }
    b.push(") ")
        .push_values(
            blocks.iter().flat_map(|block| {
                block
                    .transactions()
                    .enumerate()
                    .map(move |(i, tx)| (block, tx, i))
            }),
            |mut b, (block, tx, tx_index)| {
                let error = block.error(tx_index);
                let height = block.header().height();

                b.push_bind(AsText(tx.hash()))
                    .push_bind(height.get() as u32)
                    .push_bind(DateTime::from_timestamp_millis(
                        tx.creation_time().as_millis() as i64,
                    ))
                    .push_bind(tx.time_to_live().map(|dur| dur.as_millis() as i64))
                    .push_bind(AsText(tx.authority().signatory()))
                    .push_bind(AsText(tx.authority().domain()))
                    .push_bind(AsText(SignatureDisplay(tx.signature().payload().clone())))
                    .push_bind(tx.nonce().map(|num| i64::from(num.get())))
                    .push_bind(Json(tx.metadata()))
                    .push_bind(error.map(Json))
                    .push_bind(match tx.instructions() {
                        Executable::Instructions(_) => "Instructions",
                        Executable::Wasm(_) => "WASM",
                    });
            },
        )
        .build()
        .execute(&mut *conn)
        .await?;

    QueryBuilder::new("insert into instructions(transaction_hash, value) ")
        .push_values(
            blocks.iter().flat_map(|block| {
                block
                    .transactions()
                    .filter_map(|tx| match tx.instructions() {
                        Executable::Instructions(isi_vec) => {
                            Some(isi_vec.iter().map(|isi| (tx.hash(), isi)))
                        }
                        Executable::Wasm(_) => None,
                    })
                    .flatten()
            }),
            |mut b, (tx_hash, value)| {
                let json = match value {
                    // https://github.com/hyperledger-iroha/iroha/issues/5305
                    InstructionBox::Log(log) => Json(json!({
                        "Log": {
                            "level": log.level(),
                            "msg": "[Message could not be extracted at this moment]"
                        }
                    })),
                    other => Json(json!(other)),
                };
                // dbg!(&(tx_hash, value));
                b.push_bind(AsText(tx_hash)).push_bind(json);
            },
        )
        .build()
        .execute(&mut *conn)
        .await?;

    if !assets_definitions.is_empty() {
        debug!("Inserting assets...");
        let mut b = QueryBuilder::new("insert into asset_definitions(");
        b.separated(", ")
            .push("name")
            .push("domain")
            .push("metadata")
            .push("mintable")
            .push("owned_by_signatory")
            .push("owned_by_domain")
            .push("logo")
            .push("type");
        b.push(") ")
            .push_values(&assets_definitions, |mut b, value| {
                b.push_bind(AsText(value.id().name()))
                    .push_bind(AsText(value.id().domain()))
                    .push_bind(Json(value.metadata()))
                    .push_bind(match value.mintable() {
                        Mintable::Not => "Not",
                        Mintable::Once => "Once",
                        Mintable::Infinitely => "Infinitely",
                    })
                    .push_bind(AsText(value.owned_by().signatory()))
                    .push_bind(AsText(value.owned_by().domain()))
                    .push_bind(value.logo().as_ref().map(AsText))
                    .push_bind(match &value.type_() {
                        AssetType::Store => "Store",
                        AssetType::Numeric(_) => "Numeric",
                    });
            })
            .build()
            .execute(&mut *conn)
            .await?;
        let mut b = QueryBuilder::new("insert into assets(");
        b.separated(", ")
            .push("definition_name")
            .push("definition_domain")
            .push("owned_by_signatory")
            .push("owned_by_domain")
            .push("value");
        b.push(") ")
            .push_values(&assets, |mut b, value| {
                b.push_bind(AsText(value.id().definition().name()))
                    .push_bind(AsText(value.id().definition().domain()))
                    .push_bind(AsText(value.id().account().signatory()))
                    .push_bind(AsText(value.id().account().domain()))
                    .push_bind(match value.value() {
                        AssetValue::Store(metadata) => Json(json!({
                            "kind": "Store",
                            "value": metadata
                        })),
                        AssetValue::Numeric(num) => Json(json!({
                            "kind": "Numeric",
                            "value": num
                        })),
                    });
            })
            .build()
            .execute(&mut *conn)
            .await?;
    }

    query("COMMIT").execute(&mut *conn).await?;
    query("PRAGMA foreign_keys=ON").execute(&mut *conn).await?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::iroha_client_wrap::ClientWrap;
    use iroha::crypto::KeyPair;
    use sqlx::sqlite::SqliteConnectOptions;
    use sqlx::{ConnectOptions, Connection};
    use tracing_subscriber::layer::SubscriberExt;
    use tracing_subscriber::util::SubscriberInitExt;

    // This is more for the development of scanning rather than for its testing
    #[ignore]
    #[tokio::test]
    async fn sandbox() {
        let key_pair: KeyPair = serde_json::from_value(serde_json::json!({
            "public_key": "ed0120B23E14F659B91736AAB980B6ADDCE4B1DB8A138AB0267E049C082A744471714E",
            "private_key": "802620E28031CC65994ADE240E32FCFD0405DF30A47BDD6ABAF76C8C3C5A4F3DE96F75"
        }))
        .unwrap();

        let client = ClientWrap::new(
            AccountId::new("wonderland".parse().unwrap(), key_pair.public_key().clone()),
            key_pair,
            "http://fujiwara.sora.org/v5/".parse().unwrap(),
        );

        tracing_subscriber::registry()
            .with(
                tracing_subscriber::EnvFilter::try_from_default_env()
                    .unwrap_or_else(|_| "iroha_explorer=debug,sqlx=debug".into()),
            )
            .with(tracing_subscriber::fmt::layer())
            .init();

        let mut conn = SqliteConnectOptions::new()
            .in_memory(true)
            // .filename("./sandbox.sqlite")
            .connect()
            .await
            .unwrap();
        scan_into(&client, &mut conn)
            .await
            .expect("should scan without errors");

        conn.close().await.unwrap();
    }
}
