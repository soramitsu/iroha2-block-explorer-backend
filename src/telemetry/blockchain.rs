use crate::repo::Repo;
use crate::telemetry::AVG_COMMIT_BLOCK_TIME_WINDOW;
use sqlx::FromRow;
use std::convert::Infallible;
use std::time::Duration;

const QUERY: &str = "\
select (select count() from blocks)                                            as block,
       (select created_at from blocks order by blocks.created_at desc limit 1) as block_created_at,
       (select count() from domains)                                           as domains,
       (select count() from accounts)                                          as accounts,
       (select count() from assets) + (select count() from nfts)               as assets,
       (select count() from transactions where error is null)                  as txs_accepted,
       (select count() from transactions where error is not null)              as txs_rejected,
       (select cast(((max(unixepoch(created_at, 'subsec')) -
                      min(unixepoch(created_at, 'subsec'))) * 1000) / count(distinct created_at) as integer)
        from (select created_at
              from blocks
              order by created_at desc
              limit ?))                                                       as avg_block_time_ms";

#[derive(Clone, Debug, FromRow)]
pub struct State {
    pub block: u32,
    pub block_created_at: chrono::DateTime<chrono::Utc>,
    pub domains: u32,
    pub accounts: u32,
    pub assets: u32,
    pub txs_accepted: u32,
    pub txs_rejected: u32,
    // pub parameter_max_block_time: Duration,
    // pub parameter_max_commit_time: Duration,
    // pub parameter_max_txs_per_block: u32,
    #[sqlx(rename = "avg_block_time_ms", try_from = "u64")]
    pub avg_block_time: DurationMillis,
}

#[derive(Copy, Clone, Debug)]
pub struct DurationMillis(pub Duration);

impl TryFrom<u64> for DurationMillis {
    type Error = Infallible;

    fn try_from(value: u64) -> Result<Self, Self::Error> {
        Ok(Self(Duration::from_millis(value)))
    }
}

impl State {
    pub async fn scan(repo: &Repo) -> Result<Self, crate::repo::Error> {
        let mut conn = repo.acquire_conn().await;
        let state: State = sqlx::query_as(QUERY)
            .bind(AVG_COMMIT_BLOCK_TIME_WINDOW as u32)
            .fetch_one(&mut *conn)
            .await?;
        Ok(state)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::repo::tests::test_repo;
    use insta::assert_debug_snapshot;

    #[tokio::test]
    async fn scan_test_repo() {
        let repo = test_repo().await;

        let state = State::scan(&repo).await.unwrap();
        assert_debug_snapshot!(state);
    }
}
