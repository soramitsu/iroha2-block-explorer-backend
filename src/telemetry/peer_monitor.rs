use crate::schema::GeoLocation;
use crate::schema::ToriiUrl;
use crate::telemetry::{AverageCommitTime, AVG_COMMIT_BLOCK_TIME_WINDOW};
use eyre::eyre;
use http::StatusCode;
use iroha::client::{ConfigGetDTO, Status};
use iroha::crypto::PublicKey;
use iroha::data_model::peer::Peer;
use iroha::data_model::Identifiable;
use reqwest::Client;
use serde::Deserialize;
use std::collections::BTreeSet;
use std::future::Future;
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::{mpsc, oneshot};
use tokio::task::JoinSet;
use tokio::time::{Instant, MissedTickBehavior};
use tracing::{info_span, instrument, Instrument};
use url::Url;

const GET_STATUS_INTERVAL: Duration = Duration::from_secs(5);
const GET_STATUS_DISCONNECT_TIMEOUT: Duration = Duration::from_secs(60);
const GET_PEERS_INTERVAL: Duration = Duration::from_secs(60);
const TELEMETRY_UNSUPPORTED_CHECK_INTERVAL: Duration = Duration::from_secs(300);
const GET_GEO_RETRY_INTERVAL: Duration = Duration::from_secs(60);
const GET_CONFIG_INIT_INTERVAL: Duration = Duration::from_secs(15);
const GET_CONFIG_MAX_INTERVAL: Duration = Duration::from_secs(120);
const GET_CONFIG_INTERVAL_MULTIPLIER: f64 = 1.67;

#[instrument(fields(%torii_url))]
pub fn run(torii_url: ToriiUrl) -> (mpsc::Receiver<Update>, impl Future<Output = ()> + Sized) {
    let (tx, rx) = mpsc::channel(128);
    let url = Arc::new(torii_url);

    let url2 = url.clone();
    let fut = async move {
        let mut set = JoinSet::new();

        set.spawn({
            let tx = tx.clone();
            let url3 = url2.clone();
            async move {
                let geo = match collect_geo(&url3).await {
                    Ok(x) => x,
                    Err(err) => {
                        tracing::error!(%err, "Failed to collect geo, quitting");
                        return;
                    }
                };
                tracing::debug!(?geo, "Collected peer geo data");
                let _: Result<_, _> = tx.send(Update::Geo(geo)).await;
            }
            .instrument(info_span!("geo_future", torii_url = %url2))
        });

        set.spawn(
            {
                let url3 = url2.clone();
                async move {
                    loop {
                        let cfg = get_config_with_retry(&url3).await;
                        tracing::debug!(?cfg, "Peer connected");
                        let _ = tx.send(Update::Connected(cfg)).await;

                        let (status_fin_tx, status_fin_rx) = oneshot::channel();
                        let mut set = JoinSet::new();
                        set.spawn({
                            let tx = tx.clone();
                            let url = url3.clone();
                            async move {
                                get_peers_periodic(&url, tx).await;
                                unreachable!("should never stop")
                            }
                        });
                        set.spawn({
                            let tx = tx.clone();
                            let url = url3.clone();
                            async move {
                                get_metrics_periodic_timeout(&url, tx).await;
                                let _: Result<_, _> = status_fin_tx.send(());
                            }
                        });
                        status_fin_rx.await.expect("sender should not be dropped");
                        tracing::warn!(
                            "Peer stopped responding with status, marking as disconnected"
                        );
                        let _ = tx.send(Update::Disconnected).await;
                    }
                }
            }
            .instrument(info_span!("peer_future", torii_url = %url2)),
        );

        while set.join_next().await.is_some() {}
        unreachable!("loop should never end");
    };

    (rx, fut)
}

async fn collect_geo(torii_url: &ToriiUrl) -> eyre::Result<GeoLocation> {
    #[derive(Deserialize, Debug)]
    #[serde(tag = "status", rename_all = "lowercase")]
    enum IpApiComResponse {
        Success(GeoLocation),
        Fail { message: String },
    }

    #[derive(thiserror::Error, Debug)]
    enum RequestError {
        #[error("Request to ip-api.com failed: {0:?}")]
        Http(#[from] reqwest::Error),
        #[error("Request to ip-api.com failed with message: {message}")]
        FailResponse { message: String },
    }

    let client = Client::new();
    let url = construct_ip_api_com_query(torii_url)?;
    let do_request = || {
        let client = client.clone();
        let url = url.clone();
        async move {
            let response: IpApiComResponse = client.get(url).send().await?.json().await?;
            match response {
                IpApiComResponse::Success(data) => Ok(data),
                IpApiComResponse::Fail { message } => Err(RequestError::FailResponse { message }),
            }
        }
    };

    let data = backoff::future::retry(
        backoff::ExponentialBackoffBuilder::new()
            .with_initial_interval(GET_GEO_RETRY_INTERVAL)
            .with_multiplier(1.0)
            .with_max_elapsed_time(None)
            .build(),
        || async {
            match do_request().await {
                Ok(data) => Ok(data),
                Err(RequestError::Http(err)) => {
                    tracing::warn!(%err, "Failed to get geo location (http error, will retry)");
                    Err(backoff::Error::transient(eyre!("{err}")))
                }
                Err(RequestError::FailResponse { message }) => {
                    tracing::error!(response = %message, "Failed to get geo location (service error, will not retry)");
                    Err(backoff::Error::permanent(eyre!("Got fail response")))
                }
            }
        },
    )
    .await?;

    Ok(data)
}

fn construct_ip_api_com_query(torii_url: &ToriiUrl) -> Result<Url, eyre::Error> {
    let Some(host) = torii_url.0.host_str() else {
        return Err(eyre!("Torii URL does not have host"));
    };
    let mut url = Url::parse("http://ip-api.com/json").expect("valid");
    url.path_segments_mut().expect("url has a base").push(host);
    url.query_pairs_mut()
        .append_pair("fields", "status,message,lat,lon,country,city");
    Ok(url)
}

async fn get_config_with_retry(torii_url: &ToriiUrl) -> ConfigGetDTO {
    let client = Client::new();
    let url = torii_url.0.join("/configuration").expect("valid url");

    let do_request = || {
        let client = client.clone();
        let url = url.clone();
        async move {
            let config: ConfigGetDTO = client.get(url).send().await?.json().await?;
            Ok::<_, reqwest::Error>(config)
        }
    };

    backoff::future::retry(
        backoff::ExponentialBackoffBuilder::new()
            .with_initial_interval(GET_CONFIG_INIT_INTERVAL)
            .with_max_interval(GET_CONFIG_MAX_INTERVAL)
            .with_multiplier(GET_CONFIG_INTERVAL_MULTIPLIER)
            .with_max_elapsed_time(None)
            .build(),
        || async {
            tracing::debug!("Trying to get configuration");
            match do_request().await {
                Ok(x) => Ok(x),
                Err(err) => {
                    tracing::warn!(?err, "Failed to get configuration");
                    Err(err)?
                }
            }
        },
    )
    .await
    .expect("there is no retry limit")
}

#[instrument(skip(tx), fields(%torii_url))]
async fn get_peers_periodic(torii_url: &ToriiUrl, tx: mpsc::Sender<Update>) -> NoReturn {
    let client = Client::new();
    let url = torii_url.0.join("/peers").expect("valid url");
    let get = || async {
        let peers: Vec<Peer> = client.get(url.clone()).send().await?.json().await?;
        Ok::<_, reqwest::Error>(peers)
    };

    let mut interval = tokio::time::interval(GET_PEERS_INTERVAL);
    interval.set_missed_tick_behavior(MissedTickBehavior::Skip);

    loop {
        tracing::trace!("Collecting peers");
        match get().await {
            Ok(peers) => {
                tracing::trace!(?peers, "Collected peers from peer");
                let _ = tx
                    .send(Update::Peers(
                        peers
                            .into_iter()
                            .map(|x| x.id().public_key().clone())
                            .collect(),
                    ))
                    .await;
            }
            Err(err) => {
                tracing::warn!(?err, "Failed to get peers from peer");
            }
        }
        interval.tick().await;
    }
}

enum NoReturn {}

#[instrument(skip(tx), fields(%torii_url))]
async fn get_metrics_periodic_timeout(torii_url: &ToriiUrl, tx: mpsc::Sender<Update>) {
    #[derive(thiserror::Error, Debug)]
    enum GetError {
        #[error("http error: {0}")]
        Http(#[from] reqwest::Error),
        #[error("telemetry is not available")]
        NotImplemented,
    }

    let mut avg_commit_time = AverageCommitTime::<AVG_COMMIT_BLOCK_TIME_WINDOW>::new();
    let client = Client::new();
    let url = torii_url.0.join("/status").expect("valid url");
    let get = || async {
        let resp = client.get(url.clone()).send().await?;
        if resp.status() == StatusCode::NOT_IMPLEMENTED {
            return Err(GetError::NotImplemented);
        }
        let status: Status = resp.json().await?;
        Ok(status)
    };

    let mut interval = tokio::time::interval(GET_STATUS_INTERVAL);
    interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
    let mut count_failure_from: Instant = Instant::now();

    loop {
        tracing::trace!("Collecting status");
        match get().await {
            Ok(status) => {
                tracing::trace!(?status, "Collected status");
                count_failure_from = Instant::now();
                let block_commit_time = Duration::from_millis(status.commit_time_ms);
                avg_commit_time.observe(status.blocks, block_commit_time);
                let metrics = Metrics {
                    // peers: status.peers as u32,
                    block: status.blocks as u32,
                    block_commit_time,
                    avg_commit_time: avg_commit_time.calculate().expect("BUG: just updated"),
                    queue_size: status.queue_size as u32,
                    uptime: status.uptime.0,
                };
                let _ = tx.send(Update::Metrics(metrics)).await;
            }
            Err(GetError::Http(err)) => {
                tracing::warn!(?err, "Failed to get status");
                let elapsed = Instant::now() - count_failure_from;
                if elapsed >= GET_STATUS_DISCONNECT_TIMEOUT {
                    tracing::warn!(%url, "Peer does not respond for too long, terminate status checks");
                    // disconnected message is sent externally
                    break;
                }
            }
            Err(GetError::NotImplemented) => {
                tracing::info!("Peer does not implement telemetry");
                let _ = tx.send(Update::TelemetryUnsupported).await;
                tokio::time::sleep(TELEMETRY_UNSUPPORTED_CHECK_INTERVAL).await;
            }
        }
        interval.tick().await;
    }
}

#[derive(Clone, Copy, Debug)]
pub struct Metrics {
    // pub peers: u32,
    pub block: u32,
    pub block_commit_time: Duration,
    pub avg_commit_time: Duration,
    pub queue_size: u32,
    pub uptime: Duration,
}

#[derive(Clone, Debug)]
pub enum Update {
    Connected(ConfigGetDTO),
    Disconnected,
    TelemetryUnsupported,
    Metrics(Metrics),
    Geo(GeoLocation),
    Peers(BTreeSet<PublicKey>),
}

#[cfg(test)]
mod tests {
    use super::*;
    use tokio::spawn;

    #[test]
    fn construct_ip_api_com_urls() {
        let result = construct_ip_api_com_query(&"http://iroha.tech/".parse().unwrap()).unwrap();
        assert_eq!(
            result.to_string(),
            "http://ip-api.com/json/iroha.tech?fields=status%2Cmessage%2Clat%2Clon%2Ccountry%2Ccity"
        );

        let result =
            construct_ip_api_com_query(&"https://fujiwara.sora.org/v5".parse().unwrap()).unwrap();
        assert_eq!(
            result.to_string(),
            "http://ip-api.com/json/fujiwara.sora.org?fields=status%2Cmessage%2Clat%2Clon%2Ccountry%2Ccity"
        );
    }

    #[tokio::test]
    #[ignore]
    async fn debug_monitor() {
        crate::init_test_logger();
        let (mut rx, fut) = run("http://localhost:8080".parse().unwrap());
        spawn(fut);

        while let Some(msg) = rx.recv().await {
            tracing::info!(?msg, "Message from monitor");
        }
    }
}
