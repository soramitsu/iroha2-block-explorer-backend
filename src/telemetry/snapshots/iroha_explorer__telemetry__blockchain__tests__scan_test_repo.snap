---
source: src/telemetry/blockchain.rs
expression: state
---
State {
    block: 17,
    block_created_at: 2025-04-03T03:18:01.509Z,
    domains: 4,
    accounts: 6,
    assets: 4,
    txs_accepted: 14,
    txs_rejected: 4,
    avg_block_time: DurationMillis(
        2.843s,
    ),
}
